### PR TITLE
iCj - Signatures json format

### PR DESCRIPTION
**Pull request by wargio**
r_json to encode structured data.
r_pkcs7 added json encoding and removed R_MOVE_PTR
r_x509 added json encoding and removed R_MOVE_PTR
r_asn1 added "binary" type

### CODE REVIEW

**Comment by radare** [comment]
Using json as unique intermediate object for all data replacing sdb and rlist can be interesting for ortogonality. But i would probably think twice before going in that direction. The pr looks good so far.

**Comment by wargio** [comment]
Mine is a simple implementation of the encoder.
It's not ordered, etc.. Improvements are welcome! :D

File: binr/rabin2/rabin2.c [hunk 1]
```diff
@@ -432,7 +432,7 @@ static int rabin_do_operation(const char *op) {
 			}
 		}
 		if (plg->signature) {
-			const char *sign = plg->signature (cur);
+			const char *sign = plg->signature (cur, rad == R_CORE_BIN_JSON);
```
Comment (wargio) [inline binr/rabin2/rabin2.c]: Not freed! dammit!

**Comment by radare** [comment]
damn i though i have sent my comments :(

File: libr/util/r_json.c [hunk 2]
```diff
@@ -0,0 +1,272 @@
...
+		break;
+	case R_JS_ARRAY:
+		for (i = 0; i < var->array.l; i++) {
+			r_json_var_free (var->array.a[i]);
+		}
+		free (var->array.a);
+		break;
+	case R_JS_OBJECT:
+		for (i = 0; i < var->object.l; i++) {
+			r_json_var_free (var->object.a[i]);
```
Comment (wargio) [inline libr/util/r_json.c]: i didn't thought about that.

**Comment by wargio** [comment]
tested via 

```
	char* c = NULL;
	RJSVar *shared = r_json_object_new ();
	RJSVar *nullobj = r_json_object_new ();
	r_json_object_add (nullobj, "array_null0", r_json_null_new ());
	r_json_object_add (nullobj, "array_null2", r_json_null_new ());
	r_json_object_add (nullobj, "array_null3", r_json_null_new ());
	RJSVar *arr = r_json_array_new (3);
	r_json_array_add (arr, r_json_string_new ("random string1"));
	r_json_array_add (arr, r_json_array_new (2));
	arr->array.a[4]->array.a[0] = r_json_boolean_new (0);
	r_json_array_add (arr, r_json_array_new (2));
	RJSVar *obj = r_json_object_new ();
	r_json_object_add (obj, "boolean", r_json_boolean_new (1));
	r_json_object_add (obj, "number", r_json_number_new (1584258));
	r_json_object_add (obj, "nullable", r_json_null_new ());
	r_json_object_add (obj, "array", r_json_array_new (0));
	r_json_object_add (obj, "array_null", arr);
	r_json_object_add (obj, "object", r_json_object_new ());
	r_json_object_add (obj, "object_null", nullobj);
	//shared
	r_json_object_add (shared, "shared_array", arr);
		
	c = r_json_stringify (obj, 0);
	printf ("compressed! %s\n", c);
	free (c);
	c = r_json_stringify (obj, 1);
	printf ("expanded! %s\n", c);
	free (c);
	r_json_var_free (obj);
	c = r_json_stringify (shared, 0);
	printf ("shared! %s\n", c);
	free (c);
	r_json_var_free (shared);
```

**Comment by wargio** [comment]
@radare appveyor build system seems broken, something wrong happened.

**Comment by radare** [comment]
Skuater has been evaluating the use of a different 3rd party json parser/generator which is portable to msvc. So maybe its good to avoid some dupped work here

**Comment by radare** [comment]
Yes my fault. Forget about appveyour for now. I'll fix that later tonight

**Comment by wargio** [comment]
so what should i do? wait till somebody joins the json parsers/encoders to a unique one?

**Comment by radare** [comment]
i think is better to merge this before we discuss about possible alternatives, otherwise it will never be merged :P

**Comment by radare** [comment]
can you evaluate the switch to just use printf like we do everywhere? @skuater can you link the jsmn library you suggested me the other day?

**Comment by wargio** [comment]
are you talking to me for the printf? if yes, then it's impossible to do that. the code would be unreadable.
i prefer to wait the encoder. anyway, jsmn is just a parser, not an encoder.

**Comment by radare** [comment]
Right, maybe it was another one. Lets keep it for now :)

**Pull request merged**